pkg/ebpf/generic_map.go

```diff
@@ -0,0 +1,327 @@
+	return g.IterateWithOptions(IteratorOptions{BatchSize: defaultBatchSize})
+}
+
+func (g *GenericMap[K, V]) valueTypeCanUseUnsafePointer() bool {
+	// Simple test for now, but we probably will need to add more cases,
+	// as I am not 100% sure of the behavior of structs with maps
+	return !g.isPerCPU() // PerCPU maps use slices, so we need to pass them directly
+}
+
+// IterateWithOptions returns an iterator for the map, which transparently chooses between batch and single item
+// iterations. This version allows choosing options
+func (g *GenericMap[K, V]) IterateWithOptions(itops IteratorOptions) GenericMapIterator[K, V] {
+	if itops.BatchSize == 0 {
+		itops.BatchSize = defaultBatchSize // Default value for batch sizes. Possibly needs more testing to find an optimal default
+	}
+	if itops.BatchSize > int(g.m.MaxEntries()) {
+		itops.BatchSize = int(g.m.MaxEntries())
+	}
+
+	if BatchAPISupported() && !g.isPerCPU() && !itops.ForceSingleItem {
+		it := &genericMapBatchIterator[K, V]{
+			m:                            g.m,
+			batchSize:                    itops.BatchSize,
+			keys:                         make([]K, itops.BatchSize),
+			values:                       make([]V, itops.BatchSize),
+			valueTypeCanUseUnsafePointer: g.valueTypeCanUseUnsafePointer(),
+		}
+
+		// Do an initial copy of the keys/values slices to avoid allocations
+		it.keysCopy = it.keys
+		it.valuesCopy = it.values
+
+		return it
+	}
+
+	return &genericMapItemIterator[K, V]{
+		it:                           g.m.Iterate(),
+		valueTypeCanUseUnsafePointer: g.valueTypeCanUseUnsafePointer(),
+	}
+}
+
+type genericMapItemIterator[K interface{}, V interface{}] struct {
+	it                           *ebpf.MapIterator
+	valueTypeCanUseUnsafePointer bool
+}
+
+func (g *genericMapItemIterator[K, V]) Next(key *K, value *V) bool {
+	// we resort to unsafe.Pointers because by doing so the underlying eBPF
+	// library avoids marshaling the key/value variables while traversing the map
+	// However, in some cases (slices, structs) we need to pass the variable directly
+	// so that the library detects the type correctly
+	if g.valueTypeCanUseUnsafePointer {
+		return g.it.Next(unsafe.Pointer(key), unsafe.Pointer(value))
+	}
+
+	return g.it.Next(unsafe.Pointer(key), value)
+}
+
+func (g *genericMapItemIterator[K, V]) Err() error {
+	return g.it.Err()
+}
+
+type genericMapBatchIterator[K interface{}, V interface{}] struct {
+	m                            *ebpf.Map
+	batchSize                    int
+	cursor                       ebpf.BatchCursor
+	keys                         []K
+	values                       []V
+	keysCopy                     any // A pointer to keys of type "any", used to avoid allocations when calling BatchLookup
+	valuesCopy                   any
+	currentBatchSize             int
+	inBatchIndex                 int
+	err                          error
+	totalCount                   int
+	lastBatch                    bool
+	valueTypeCanUseUnsafePointer bool
```

Comment:
I think it is critical to have documentation for the fields here, we have 13 fields and it is unclear what's the usage and meaning of each of them